src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -7015,20 +7015,20 @@ private enum ParseTypeMode
             FirstElementOfPossibleTupleLiteral,
         }
 
-        private TypeSyntax ParseType(ParseTypeMode mode = ParseTypeMode.Normal)
+        private TypeSyntax ParseType(ParseTypeMode mode = ParseTypeMode.Normal, bool inSwitchArmPattern = false)
```

Comment:
`whenIsKeyword` got renamed to `inSwitchArmPattern` to try to be clearer.  We pass 'true' in here when parsign the pattern for a switch-statement or switch-expression pattern.   And it controls how we view 'where' and a few other error recovery rules (including new ones in this pr).